docs/pnpm.md

```diff
@@ -0,0 +1,282 @@
+This should include the `pnpm-workspace.yaml` file as well as all `package.json` files
+in the pnpm workspace.
+The pnpm lock file update will fail if `data` is missing any files required to run
+`pnpm install --lockfile-only` or `pnpm import`.
+
+> To list all local `package.json` files that pnpm needs to read, you can run
+> `pnpm recursive ls --depth -1 --porcelain`.
+
+A `.aspect/rules/external_repository_action_cache/npm_translate_lock_<hash>` file will be created
+and used to determine when the `pnpm-lock.yaml` file should be updated.
+This file should be checked into the source control along with the `pnpm-lock.yaml` file.
+
+When the `pnpm-lock.yaml` file needs updating, `npm_translate_lock` will automatically:
+
+-   run `pnpm import` if there is a `npm_package_lock` or `yarn_lock` attribute specified.
+-   run `pnpm install --lockfile-only` otherwise.
+
+To update the `pnpm-lock.yaml` file manually, either
+
+-   [install pnpm](https://pnpm.io/installation) and run `pnpm install --lockfile-only` or `pnpm import`
+-   use the Bazel-managed pnpm by running `bazel run -- @pnpm//:pnpm --dir $PWD install --lockfile-only` or `bazel run -- @pnpm//:pnpm --dir $PWD import`
+
+If the `ASPECT_RULES_JS_FROZEN_PNPM_LOCK` environment variable is set and `update_pnpm_lock` is True,
+the build will fail if the pnpm lock file needs updating.
+It is recommended to set this environment variable on CI when `update_pnpm_lock` is True.
+
+## Working with packages
+
+### Patching
+
+You can apply patches to packages you fetch remotely such as from npm.
+Use the `patches` and `patch_args` attributes of `npm_translate_lock`.
+These are designed to be similar to the same-named attributes of
+[http_archive](https://bazel.build/rules/lib/repo/http#http_archive-patch_args).
+
+Paths in patch files must be relative to the root of the package.
+If the version is left out of the package name, the patch will be applied to every version of the npm package.
+
+`patch_args` defaults to `-p0`, but `-p1` will usually be needed for patches generated by git.
+
+In case multiple entries in `patches` match, the list of patches are additive.
+(More specific matches are appended to previous matches.)
+However if multiple entries in `patch_args` match, then the more specific name matches take precedence.
+
+For example,
+
+```starlark
+npm_translate_lock (
+    ...
+    patches = {
+        "@foo/bar": ["//:patches/foo+bar.patch"],
+        "fum@0.0.1": ["//:patches/fum@0.0.1.patch"],
+    },
+    patch_args = {
+        "*": ["-p1"]
+        "@foo/bar": ["-p0"]
+        "fum@0.0.1": ["-p2"]
+    },
+)
+```
+
+### Lifecycles
+
+npm packages have "lifecycle scripts" such as `postinstall` which are documented here:
+<https://docs.npmjs.com/cli/v9/using-npm/scripts#life-cycle-scripts>
+
+We refer to these as "lifecycle hooks".
+You can disable this feature completely by setting `lifecycle_hooks = False` in `npm_translate_lock`.
```

Comment:
Full disable is done by `run_lifecycle_hooks = False` or by
```
lifecycle_hooks = { "*": [] }
```

Reply:
run_lifecycle_hooks is deprecated so I'll use the second one